astropy/io/fits/file.py

```diff
@@ -119,10 +120,24 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
         if mode not in IO_FITS_MODES:
             raise ValueError("Mode '{}' not recognized".format(mode))
 
+        # Handle raw URLs
         if (isinstance(fileobj, string_types) and
-            mode not in ('ostream', 'append') and
-            _is_url(fileobj)):  # This is an URL.
+            mode not in ('ostream', 'append', 'update') and _is_url(fileobj)):
             self.name = download_file(fileobj, cache=cache)
+        # Handle HTTP requests (i.e. no response received yet)
+        elif isinstance(fileobj, urllib.request.Request):
+            if mode in ('ostream', 'append', 'update'):
+                raise ValueError(
+                    "Mode {} not supported for URL request".format(mode))
+            fileobj = fileobj.get_full_url()
+            self.name = download_file(fileobj, cache=cache)
+        # Handle responses from URL requests that have already been opened
+        elif isinstance(fileobj, http_client.HTTPResponse):
+            if mode in ('ostream', 'append', 'update'):
+                raise ValueError(
+                    "Mode {} not supported for HTTPResponse".format(mode))
+            from io import BytesIO
```

Comment:
The reason It's not at the top is because you can avoid the import in the failure case. I'm not sure it matters much either way.